std_execution.bs

```diff
@@ -3990,6 +3994,161 @@ from the operation before the operation completes.
+
+    * The object to which <code><i>REC</i>(<i>o</i>)</code> refers remains valid for the lifetime of the object to which <code><i>o</i></code> refers.
+
+    * The type <code><i>run-loop-opstate</i>&lt;R></code> overrides <code><i>run-loop-opstate-base</i>::execute()</code> such that <code><i>o</i>.execute()</code> is equivalent to the following:
+
+        <pre highlight="c++">
+        try {
+          if (execution::get_stop_token(<i>REC</i>(<i>o</i>)).stop_requested()) {
+            execution::set_done(std::move(<i>REC</i>(<i>o</i>)));
+          } else {
+            execution::set_value(std::move(<i>REC</i>(<i>o</i>)));
+          }
+        } catch(...) {
+          execution::set_error(std::move(<i>REC</i>(<i>o</i>)), current_exception());
+        }
+        </pre>
+
+    * The expression <code>execution::start(<i>o</i>)</code> is expression-equivalent to <code><i>o</i>.start()</code>.
+
+#### Constructor <b>[execution.contexts.run_loop.ctor]</b> #### {#spec-execution.contexts.run_loop.ctor}
+
+    <pre highlight="c++">
+    run_loop::run_loop() noexcept;
+    </pre>
+
+    1. <i>Postconditions:</i> <i>count</i> is `0` and <i>state</i> is <i>not-finishing</i>.
+
+#### Member functions <b>[execution.contexts.run_loop.members]</b> #### {#spec-execution.contexts.run_loop.members}
+
+    <pre highlight="c++">
+    <i>run-loop-opstate-base</i>* run_loop::pop_front() noexcept;
+    </pre>
+
+    1. <i>Effects:</i> Waits until one of the following conditions is `true`:
+
+        * <i>count</i> is `0` and <i>state</i> is <i>finishing</i>, in which case `pop_front()` returns `nullptr`; or
+
+        * <i>count</i> is greater than `0`, in which case `pop_front()` removes an item from the front of the queue, decrements <i>count</i> by `1`, and returns the item.
+
+    2. <i>Synchronization:</i> This function is thread-safe with respect to `push_back()` and `finish()`. Removing a queue item and decrementing <i>count</i> are executed atomically.
+
+    <pre highlight="c++">
+    void run_loop::push_back(<i>run-loop-opstate-base</i>* item) noexcept;
+    </pre>
+
+    1. <i>Effects:</i> Adds `item` to the back of the queue and increments <i>count</i> by `1`.
+
+    2. <i>Synchronization:</i> This function is thread-safe with respect to `pop_front()` and `finish()`. Adding a queue item and incrementing the <i>count</i> are executed atomically.
+
+    <pre highlight="c++">
+    <i>run-loop-scheduler</i> run_loop::get_scheduler();
+    </pre>
+
+    1. <i>Returns:</i> an instance of <code><i>run-loop-scheduler</i></code> that can be used to schedule work onto this `run_loop` instance.
+
+    <pre highlight="c++">
+    void run_loop::run();
+    </pre>
+
+    1. <i>Effects:</i> Equivalent to:
+
+        <pre highlight="c++">
+        while (auto* op = pop_front()) {
+          op->execute();
+        }
+        </pre>
+
+    2. <i>Synchronization:</i> None.
```

Comment:
FYI - this is nice from a wording perspective, but threw me off track because I expected finish/pop_front/push_back to all say that they were synchronizing with run as well.. I deleted those comments after I realized how run() was excluded from sync.

Reply:
Yes, it's important that executing the work isn't done while holding a lock that will prevent more work from being added.